metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala

```diff
@@ -64,31 +71,87 @@ final class BuildTools(
       }
     }
   }
-  def isMill: Boolean = workspace.resolve("build.sc").isFile
-  def isScalaCli: Boolean =
-    ScalaCliBuildTool
-      .pathsToScalaCliBsp(workspace)
-      .exists(_.isFile) || workspace.resolve("project.scala").isFile
-  def isGradle: Boolean = {
+  def isSbt: Boolean = sbtProject.isDefined
+  def millProject: Option[AbsolutePath] = searchForBuildTool(
+    _.resolve("build.sc").isFile
+  )
+  def isMill: Boolean = millProject.isDefined
+  def scalaCliProject: Option[AbsolutePath] =
+    searchForBuildTool(_.resolve("project.scala").isFile)
+      .orElse {
+        ScalaCliBspScope.scalaCliBspRoot(workspace) match {
+          case Nil => None
+          case path :: Nil if path.isFile => Some(path.parent)
+          case path :: Nil =>
+            scribe.info(s"path: $path")
+            Some(path)
+          case _ => Some(workspace)
+        }
+      }
+
+  def gradleProject: Option[AbsolutePath] = {
     val defaultGradlePaths = List(
       "settings.gradle",
       "settings.gradle.kts",
       "build.gradle",
       "build.gradle.kts",
     )
-    defaultGradlePaths.exists(workspace.resolve(_).isFile)
+    searchForBuildTool(root =>
+      defaultGradlePaths.exists(root.resolve(_).isFile)
+    )
   }
-  def isMaven: Boolean = workspace.resolve("pom.xml").isFile
-  def isPants: Boolean = workspace.resolve("pants.ini").isFile
-  def isBazel: Boolean = workspace.resolve("WORKSPACE").isFile
+  def isGradle: Boolean = gradleProject.isDefined
+  def mavenProject: Option[AbsolutePath] = searchForBuildTool(
+    _.resolve("pom.xml").isFile
+  )
+  def isMaven: Boolean = mavenProject.isDefined
+  def pantsProject: Option[AbsolutePath] = searchForBuildTool(
+    _.resolve("pants.ini").isFile
+  )
+  def isPants: Boolean = pantsProject.isDefined
+  def bazelProject: Option[AbsolutePath] = searchForBuildTool(
+    _.resolve("WORKSPACE").isFile
+  )
+  def isBazel: Boolean = bazelProject.isDefined
 
+  private def searchForBuildTool(
+      isProjectRoot: AbsolutePath => Boolean
+  ): Option[AbsolutePath] = {
+    projectRoot.get() match {
+      case Some(root) => if (isProjectRoot(root)) Some(root) else None
+      case None =>
+        def recIsProjectRoot(
+            path: AbsolutePath,
+            level: Int = 0,
+        ): Option[AbsolutePath] =
+          // we skip `.scala-build` and `project`, which both contain .bloop
+          if (
+            path.isDirectory && !path.toNIO.filename.startsWith(
+              "."
+            ) && path.toNIO.filename != "project"
```

Comment:
Because in "project" we have `.bloop` for meta build and we don't want to detected it. Especially if something goes wrong and we have `.bloop` only in "project". I guess we can do it differently avoid going deeper if there is some top-level build tool (I feel like there was an argument against that approach but I can't remember now).

Reply:
We wouldn't go to `project/.bloop` since we would already have `.bloop` in the main dir, no?

Reply:
In the happy path, yes. But if something goes wrong and `.bloop` in main directory isn't created, the one in `project/` still will be (we even have a test for this scenario). Connecting with meta build in that scenerio would be extremely confusing for the user.

Reply:
I changed that do discovering bloop is correlated with the chosen build tool.